### PR TITLE
Use multiple threads to generate navmeshes

### PR DESCRIPTION
**Pull request by slipher**
Some improvements compared to #2944, in particular:
- keeps code for generating on main thread only for sgame background navgen
- can cancel navgen if the sgame shuts down so as not to hang the server
- fixes regression that map-level errors are no longer cached

Also I added menu options.

### CODE REVIEW

File: src/shared/navgen/nav.cpp
Comment (DolceTriade) [inline src/shared/navgen/nav.cpp:69]: Does this need to be cleared at the end of the function? If not, please comment.

File: src/shared/navgen/nav.cpp
Comment (DolceTriade) [inline src/shared/navgen/nav.cpp:86]: Can we leave comments about which functions are main thread only here too?

File: src/shared/navgen/nav.cpp
Comment (DolceTriade) [inline src/shared/navgen/nav.cpp:1058]: Just init once outside the main loop to make the intent clear?

**Comment by DolceTriade** [comment]
What are the numbers with and without patch for:
 - Anthill
 - cube
 - plat23

I'll help test. Let's see if we can get this into 0.55

**Comment by slipher** [comment]
All right, I tested like this `time( ./daemonded.exe -set vm.sgame.type 1 -pakpath /c/unv/Unvanquished/pkg (-set g_bot_navgen_maxThreads 7) +devmap plat23 +navgen all +quit )` for blocking navgen and this `time( ./daemonded.exe -set vm.sgame.type 1 -pakpath /c/unv/Unvanquished/pkg -set g_bot_navgen_maxThreads 7 -set logs.level.sgame.navgen debug +devmap plat23 +bot fill 5 )` (plus manually clicking X at the end, so less accurate) for background navgen. Tested with NaCl on MinGW. I have a 4 cores with hyperthreading processor.

Branch | Navgen Type | plat23 | cube | anthill
--- | --- | --- | --- | ---
master | blocking navgen | 0m4.614s | 1m18.648s | 3m47.651s
master | background navgen | 0m6.513s | 1m20.323s | N/A
#2955 | blocking navgen (7 threads) | 0m2.399s | 0m26.809s | 1m14.501s
#2955 | background navgen (7 threads) | 0m2.657s | 0m29.411s | 1m18.005s


Also experimented a bit to see if the hyperthreading works well or not, I think not really. For example in a 4-threaded Anthill run I got 51542ms and 14083ms for Tyrant while with 7 threads I got 38567ms and  9689ms for Tyrant.

File: src/cgame/cg_main.cpp
Comment (DolceTriade) [inline src/cgame/cg_main.cpp:161]: consider using the max threads possible by default. no reason not to.

**Review by DolceTriade — COMMENTED** [review]
Also I need to get around to testing this on a shit VPS which only has a single core. tbh, I don't think we should expect people to run these under nacl, but we'll solicit feedback from others too about this situation regardless.

**Comment by DolceTriade** [comment]
> Also experimented a bit to see if the hyperthreading works well or not, I think not really. For example in a 4-threaded Anthill run I got 51542ms and 14083ms for Tyrant while with 7 threads I got 38567ms and 9689ms for Tyrant.

when I get back home, I'll try running these numbers on my system since I have enough cores to run all the classes in parallel. For me, anthill took ~30s. I don't expect any regressions, but let's be double sure about what we're doing here.

File: src/cgame/cg_main.cpp
Comment (DolceTriade) [reply src/cgame/cg_main.cpp:161]: https://en.cppreference.com/w/cpp/thread/thread/hardware_concurrency

File: src/shared/navgen/nav.cpp
Comment (slipher) [reply src/shared/navgen/nav.cpp:69]: The idea is to only call it once. I think there is a comment elsewhere about that but I went ahead and added a `clear()` just to avoid suspicious appearances.

File: src/shared/navgen/nav.cpp
Comment (slipher) [reply src/shared/navgen/nav.cpp:88]: I added comments to functions which can be called from a non-main thread since there are much fewer of those and they are the ones with restrictions.

File: src/shared/navgen/nav.cpp
Comment (slipher) [reply src/shared/navgen/nav.cpp]: Done.

File: src/cgame/cg_main.cpp
Comment (slipher) [reply src/cgame/cg_main.cpp:162]: Wow it even works in Nacl. I put the default on hardware_concurrency() - 1 since on Windows at least that's the most you can use without making other apps sluggish.

For the server I am leaving it at 1 background thread by default to respect other services which may be running.

Comment (slipher) [reply src/cgame/cg_main.cpp:162]: We should test that hardware_concurrency works on a Linux and a Mac system though

File: src/shared/navgen/nav.cpp
Comment (DolceTriade) [inline src/shared/navgen/nav.cpp:1063]: loadmap twice?

**Comment by DolceTriade** [comment]
So this looks good to me. I tested on my shit VPS and it seems to work well. We really shouldn't expect people to generate huge maps like anthill on their single core vps.

I'll solicit some more feedback but otherwise and test perhaps on some more maps, but then we should merge this prior to 0.55.

File: src/cgame/cg_main.cpp
Comment (slipher) [reply src/cgame/cg_main.cpp:162]: Tested that this API works inside and outside Nacl on Linux and Mac.

**Review by DolceTriade — APPROVED** [review]

**Comment by DolceTriade** [comment]
LGTM. Let's wait for sweet to give this a try and then we can merge.

**Comment by sweet235** [comment]
My server does not have the means to test this. But this is about local games mostly, and you already tested this.

**Comment by DolceTriade** [comment]
Cool, looks ready to merge.

**Comment by sweet235** [comment]
I am still not sure if adding this complexity is worth it. Even if the code is free of bugs, it will make the server a much more dangerous thing to maintain for any sysadmin. A wrong command might spawn enough threads to grind the whole machine to a halt.

**Comment by DolceTriade** [comment]
> I am still not sure if adding this complexity is worth it. Even if the code is free of bugs, it will make the server a much more dangerous thing to maintain for any sysadmin. A wrong command might spawn enough threads to grind the whole machine to a halt.

I think that this is a really cool feature that significantly reduces the time it takes to generate navmeshes. I think that there are reasonable safeguards to prevent server stoppages.

**Comment by sweet235** [comment]
> I think that there are reasonable safeguards to prevent server stoppages.

I'm good, my server does nothing else. But once we merge this, people will probably have to put our server in a managed container on machines that do other work too.

**Pull request merged**
